### PR TITLE
applied isort to seldon_core and tests

### PR DESCRIPTION
**Pull request by zyxue**
**What this PR does / why we need it**:

Sorted python imports in a principled way with [isort](https://pycqa.github.io/isort/), and sorted imports improve code readability.

**Which issue(s) this PR fixes**:

NA

**Special notes for your reviewer**:

what's done includes:

* added `isort` to `requirements-dev.txt`
* configed `isort` in `pyproject.toml` for compatibility with black (see https://pycqa.github.io/isort/docs/configuration/black_compatibility/)
* and the run `isort tests seldon_core`

**Does this PR introduce a user-facing change?**:

No, this PR shouldn't change any code logic

### CODE REVIEW

**Comment by seldondev** [comment]
Hi @zyxue. Thanks for your PR.

I'm waiting for a [SeldonIO](https://github.com/orgs/SeldonIO/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/SeldonIO/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository.
</details>

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **rafalskolasinski**
You can assign the PR to them by writing `/assign @rafalskolasinski` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: python/seldon_core/flask_utils.py
Comment (RafalSkolasinski) [inline python/seldon_core/flask_utils.py:8]: Does this need to be split into separate lines? 
If yes, I'd prefer to use `( ... )` like we have in other places.

**Comment by seldondev** [comment]
Thu Feb  4 17:34:05 UTC 2021
The logs for [integration] [1] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2927/1.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2927 --build=1

File: python/seldon_core/flask_utils.py
Comment (zyxue) [reply python/seldon_core/flask_utils.py:8]: I see. it's due to the aliasing of `jsonify as flask_jsonify`. Otherwise isort would put them into one line like
```
from flask import current_app, jsonify, request
```

Comment (RafalSkolasinski) [reply python/seldon_core/flask_utils.py:8]: Oh, I didn't realise we had aliasing in one line :D
I guess the best is to have then
```
from flask import current_app, request
from flask import jsonify as flask_jsonify
```
?

Comment (zyxue) [reply python/seldon_core/flask_utils.py:8]: it's because `request` comes after `jsonify` alphabetically. if it's like

```
from flask import ctx, current_app
from flask import jsonify as flask_jsonify
```
isort would keep `ctx` and `current_app` in one line.

Comment (RafalSkolasinski) [reply python/seldon_core/flask_utils.py:8]: What if we use `# isort:skip`:
```
from flask import current_app, request
from flask import jsonify as flask_jsonify   # isort:skip
```
?

Comment (zyxue) [reply python/seldon_core/flask_utils.py:8]: I think the main benefit of isort is that it has a set of rules for ordering imports and consistent. Perhaps we could customize then, but then it feels ad-hoc just like the current heuristic way unless we have a good reason like to be compatible with black.

I see it this way, like black, it is opinioned but consistent.

Comment (zyxue) [reply python/seldon_core/flask_utils.py:8]: tried, I didn't know this, but isort would put a line in between
```
from flask import current_app, request

from flask import jsonify as flask_jsonify   # isort:skip
```

Comment (RafalSkolasinski) [reply python/seldon_core/flask_utils.py:8]: Ok, cool, let's leave it like that

**Comment by RafalSkolasinski** [comment]
BTW. I think we should also update the `fmt` or `lint` target in the `python/Makefile`. Not sure which one is better, @adriangonz? 
Also we could include installation of isort in the `setup_linter` target in the same Makefile.

**Comment by adriangonz** [comment]
That would make a lot of sense @RafalSkolasinski. Regarding lint vs fmt, my understanding is that lint only says whether is OK or not (but doesn't modify anything), whereas fmt does modify files.

For the case of isort, we could add it on lint, to flag if the imports haven't been sorted (useful for CI). And we could add it on fmt to actually sort the imports on the file.

Does that makes sense? What are your thoughts?

**Comment by RafalSkolasinski** [comment]
Right, that's the proper distinction! 

Sounds good, just need to keep in mind then that adding it on `lint` would need to be in form of a check that if changes are made it should fail.

**Comment by adriangonz** [comment]
@RafalSkolasinski I think isort already has a --check-only option just for that. These tools tend to include similar options to use as linters.

**Comment by RafalSkolasinski** [comment]
Perfecto!

@zyxue, once applying the suggestions above and passing the integration tests we will be good to merge

**Comment by zyxue** [comment]
thanks for the clarification. when is

**Comment by zyxue** [comment]
my bad, I misunderstood. please ignore my last comment

**Comment by zyxue** [comment]
Why did black skip `testing/scripts`?

**Comment by RafalSkolasinski** [comment]
it should only be skipping `testing/scripts/proto` there

File: python/Makefile
Comment (RafalSkolasinski) [inline python/Makefile:103]: Should be `testing/scripts`  not `testing|scripts`

**Review by RafalSkolasinski — CHANGES_REQUESTED** [review]

File: python/Makefile
Comment (zyxue) [reply python/Makefile:103]: sorry for confusion, fixed.

Comment (RafalSkolasinski) [reply python/Makefile]: Hmm... actually, maybe just `scripts` is enough because this applies when being inside `../testing` already.

**Review by RafalSkolasinski — CHANGES_REQUESTED** [review]

**Comment by zyxue** [comment]
@RafalSkolasinski , rerun `make lint` changes these files:
```
	modified:   licenses/license.txt
	modified:   licenses/license_info.csv
	modified:   licenses/license_info.no_versions.csv
```
is that expected, should I commit these changes?

**Comment by RafalSkolasinski** [comment]
Yeah, it probably adds `isort` licence as this is new deps.

**Comment by zyxue** [comment]
I don't see isort in the linceses change though

**Comment by zyxue** [comment]
@RafalSkolasinski , all tests pass now, please let me know if there is more to update.

**Comment by seldondev** [comment]
Thu Feb  4 22:51:35 UTC 2021
The logs for [integration] [3] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2927/3.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2927 --build=3

**Comment by seldondev** [comment]
Thu Feb  4 22:51:35 UTC 2021
The logs for [notebooks] [2] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2927/2.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2927 --build=2

File: python/Makefile
Comment (RafalSkolasinski) [inline python/Makefile:103]: Note sure about this one, I think there was a reason why why excluded `testing/scripts`.
@axsaucedo @adriangonz?

Comment (adriangonz) [reply python/Makefile:103]: @RafalSkolasinski we exclude the `proto` folder within `testing/scripts`. It would be good to double check whether that expression also captures that or not. Otherwise `black` is gonna go crazy formatting the generating code for the protobufs.

Comment (RafalSkolasinski) [reply python/Makefile:103]: Yeah, make total sense for `protos`. Was just wondering if there is something else in `testing/scripts` we would like to leave unmoddified.

File: python/seldon_core/proto/tensorflow/core/framework/__init__.py
Comment (RafalSkolasinski) [inline python/seldon_core/proto/tensorflow/core/framework/__init__.py:5]: We want to exclude protos also from `isort`.

**Review by RafalSkolasinski — CHANGES_REQUESTED** [review]

File: python/Makefile
Comment (RafalSkolasinski) [reply python/Makefile:103]: Okay, cool, I don't see any black re-formats in the protos. Only `isort` touched some imports, added separate comment for that: https://github.com/SeldonIO/seldon-core/pull/2927#discussion_r570865145

File: python/seldon_core/flask_utils.py
Comment (RafalSkolasinski) [reply python/seldon_core/flask_utils.py:8]: And by that I meant with ` # isort:skip` ; - )

**Review by RafalSkolasinski — CHANGES_REQUESTED** [review]

File: python/seldon_core/proto/tensorflow/core/framework/__init__.py
Comment (zyxue) [reply python/seldon_core/proto/tensorflow/core/framework/__init__.py:5]: This isn't about formatting pb2 files per se, it just formats how they're formatted

File: python/Makefile
Comment (zyxue) [reply python/Makefile:103]: I was thinking formatting pb2 files would be fine too. I don't see any disadvantage with that as long as it's always formatted whenever they're regenerated. But I'll follow the convention and ignore them here.

File: python/seldon_core/proto/tensorflow/core/framework/__init__.py
Comment (zyxue) [reply python/seldon_core/proto/tensorflow/core/framework/__init__.py:5]: Hmm. But this is inside the proto for, let me see

Comment (RafalSkolasinski) [reply python/seldon_core/proto/tensorflow/core/framework/__init__.py:5]: > This isn't about formatting pb2 files per se, it just formats how they're formatted

Now that's a definitely a bigger change that should go into separate issue for discussion ; - )

Comment (zyxue) [reply python/seldon_core/proto/tensorflow/core/framework/__init__.py:5]: Sorry I meant how they're **imported***.

Ok, I agree this issue is getting overly large, I'll limit the change to seldon_core only.

Comment (zyxue) [reply python/seldon_core/proto/tensorflow/core/framework/__init__.py:5]: I think the principal is that every manually written python files (except for auto generated ones e.g pb2 files) should follow consistent formatting enforced by black and isort, is that agreeable?

Comment (RafalSkolasinski) [reply python/seldon_core/proto/tensorflow/core/framework/__init__.py:5]: Yes, that's right. What we write manually should be under `black` and `isort`. So all python codes except those under `proto` should fall into it. 

The tensorflow protos I think I coming from 3rd party library so we should not format them.

Comment (zyxue) [reply python/seldon_core/proto/tensorflow/core/framework/__init__.py]: I see. I probably committed this one by mistake as `proto` directory has been indeed ignored. I've changed it back.

File: python/Makefile
Comment (zyxue) [reply python/Makefile:103]: https://github.com/SeldonIO/seldon-core/tree/master/testing/scripts/proto is empty now except for a `.keep` file ...

**Comment by zyxue** [comment]
due to conflicts, I'm rebasing to master

**Comment by zyxue** [comment]
Curious, from https://github.com/SeldonIO/seldon-core/pull/2769, I noticed that you didn't squash all commits during merge:
```
git --no-pager show --oneline --no-patch 35d9c8..f89aa4
f89aa4043 newline
f380dac57 added venv/ to .gitignore
e39767883 corrected typo
8fd5e5a29 reblacked code in seldon_core
57e6ddd32 updated black to black 20.8b1 in setup_linter
```
Is there a reason for it? I thought squashing all these commits into just one would make the master history cleaner.

**Comment by RafalSkolasinski** [comment]
> Is there a reason for it? I thought squashing all these commits into just one would make the master history cleaner.

I don't think we are very strict on that. Also merging happens automatically by the `seldondev` bot.

**Comment by zyxue** [comment]
> > Is there a reason for it? I thought squashing all these commits into just one would make the master history cleaner.
> 
> I don't think we are very strict on that. Also merging happens automatically by the `seldondev` bot.

ah, I see. Thanks.

I'm done with my changes now.

**Comment by seldondev** [comment]
Fri Feb  5 16:31:49 UTC 2021
The logs for [notebooks] [4] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2927/4.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2927 --build=4

**Comment by seldondev** [comment]
Fri Feb  5 16:31:57 UTC 2021
The logs for [integration] [5] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2927/5.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2927 --build=5

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *[RafalSkolasinski](https://github.com/SeldonIO/seldon-core/pull/2927#issuecomment-774143503 "Approved")*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com#issuecomment-774143503/SeldonIO/seldon-core/blob/master/OWNERS)~~ [RafalSkolasinski]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by zyxue** [comment]
@RafalSkolasinski , seems blocked by `Pending — Not mergeable. Should not have needs-ok-to-test label.`

**Pull request merged**
